common/setups/rasr/nn_system.py

```diff
@@ -0,0 +1,554 @@
+import i6_core.meta as meta
+import i6_core.rasr as rasr
+import i6_core.returnn as returnn
+import i6_core.util as util
+
+from .rasr_system import RasrSystem
+
+from .util import (
+    RasrInitArgs,
+    ReturnnRasrDataInput,
+    OggZipHdfDataInput,
+    NnArgs,
+    RasrSteps,
+)
+
+from i6_private.users.luescher.util import (
+    load_multipath_from_bundle,
+    load_task_dependent_data,
+)
+
+# -------------------- Init --------------------
+
+Path = tk.setup_path(__package__)
+
+# -------------------- System --------------------
+
+
+class NnSystem(RasrSystem):
+    """
+    - 5 corpora types: train, devtrain, cv, dev and test
+        devtrain is a small split from the train set which is evaluated like
+        the cv but not used for error calculating. Since we can have different
+        datasubsets per subepoch, we do not caculate the tran score/error on
+        a consistent datasubset
+    - two training data settings: defined in returnn config or not
+    - 3 different types of decoding: returnn, rasr, rasr-label-sync
+    - 2 different lm: count, neural
+    - cv is dev for returnn training
+    - dev for lm param tuning
+    - test corpora for final eval
+
+    settings needed:
+    - am
+    - lm
+    - lexicon
+    - ce training
+    - ce recognition
+    - ce rescoring
+    - smbr training
+    - smbr recognition
+    - smbr rescoring
+    """
+
+    def __init__(
+        self,
+        returnn_root: Optional[str] = None,
+        returnn_python_home: Optional[str] = None,
+        returnn_python_exe: Optional[str] = None,
+    ):
+        super().__init__()
+
+        self.returnn_root = returnn_root or gs.RETURNN_ROOT
+        self.returnn_python_home = returnn_python_home or gs.RETURNN_PYTHON_HOME
+        self.return_python_exe = returnn_python_exe or gs.RETURNN_PYTHON_EXE
+
+        self.tf_fwd_input_name = "tf-fwd-input"
+
+        self.devtrain_corpora = []
+        self.cv_corpora = []
+
+        self.train_input_data = None
+        self.cv_input_data = None
+        self.devtrain_input_data = None
+        self.dev_input_data = None
+        self.test_input_data = None
+
+        self.train_cv_pairing = None
+
+        self.datasets = {}
+
+        self.oggzips = {}
+
+        self.nn_configs = {}
+        self.nn_models = {}
+
+    # -------------------- Helpers --------------------
+    @staticmethod
+    def adapt_returnn_config_for_recog(returnn_config: returnn.ReturnnConfig):
+        """
+        Adapt a RETURNN config for recognition, e.g., remove loss and use log softmax activation in last layer
+
+        :param ReturnnConfig config:
+        :rtype ReturnnConfig:
+        """
+        assert isinstance(returnn_config, returnn.ReturnnConfig)
+        config = copy.deepcopy(returnn_config)
+        forward_output_layer = config.config.get("forward_output_layer", "output")
+        network = config.config.get("network")
+        for layer_name, layer in network.items():
```

Comment:
I agree. Based on comments by Nick. I ll wait to see his solution which should hopefully mean that we can remove this function

Reply:
Okay, so we can not completely drop this function I think, but there are some things that can be removed.

 - Targets do not to be dropped in RETURNN, for me I can always keep them.
 - I would consider using lstmp an user error, and not something to fix here
 - For the log outputs, they can be part of the network in the first place, and then we can tell the tf_flow helper what output to use. This would be more consistent and flexible (and safe heuristics code).
 
 For the "cropped" thing I would refer to @michelwi solution, as this is way too specific to be included here.